### DISCUSSION TITLE
How do I avoid pendingComponent being shown again after first load?

### DISCUSSION THREAD

**Discussion by brunoti** [TanStack/router#2706] [Q&A]
Let's say I have a route, with a loader that depends on a search param called search.

This search param is updated based on user input for searching data inside the router. After first load I don't want to show pending component again, as it would rerender everything. I want to be able to keep the user at the same place and not call the loader again but now only the search params.

Is that possible?

**Comment by rostero1** [comment]
I wonder if this is a bug instead of a feature

**Comment by rostero1** [comment]
It appears that the pendingComponent will display when navigating to the same page and updating just the parameters only when the loaderDeps is used.  I'm not sure why that is.  It seems that the loader code is still called whenever the loaderDeps option isn't run.  

Here's a demo: https://stackblitz.com/edit/github-nnbjygkg-wci72ac4?file=src%2Fmain.tsx

**Reply by rostero1** [reply]
With the Deps tab, I'm also I'm curious why there is such a delay after you hit Next between when "Pending: Yes" is shown and the pendingComponent is rendered

**Reply by rostero1** [reply]
Maybe related: https://github.com/TanStack/router/issues/2990

**Reply by rostero1** [reply]
Discussion in discord:

```
createRouter({
   routeTree,
   defaultRemountDeps: ({search, params}) => params,
})
```

**Comment by arackaf** [comment]
Have you tried setting `pendingMs` ? I *think* that delays the showing of the pending component for X ms, on subsequent navigations. But I believe it's ignored for the initial page load (the pending component is showed immediately, when first rendering the web app).

I *think* - obv confirm :)

**Reply by rostero1** [reply]
Thanks.  That got rid of the initial delay between hitting next and it loading.

**Comment by arackaf** [comment]
Honestly I think you might want to “graduate” to react-query. Use the loader to prefetch whatever queries your route uses

As search params change, update the queries. 

You could even still use the pendingComponent if you use useSuspenseQuery, and to avoid existing content from suspending when search params change you could use useDeferredValue (iirc, I’m afk) to feed your query params.

Examples here https://frontendmasters.com/blog/tanstack-router-data-loading-2/

**Reply by rostero1** [reply]
I think this is the way to go.  Thanks for the article.  I was able to write a demo that met my needs.

**Comment by akrynski-1** [comment]
I think pending component should not be shown on subsequent navigations to the same route.
It makes user experience worse.
Increasing `pendingMs` is not a good solution either, it would delay showing pending component on first navigation.
Using `useSuspenseQuery()` is even worse because it shows fallback immediately.

**Comment by shx-riff** [comment]
I use this workaround:
```
export const Route = createFileRoute('/users')({
  validateSearch: z.object({
    page: fallback(z.number().default(1), 1),
    perPage: fallback(z.number().default(15), 15),
  }),
  loaderDeps: ({ search }) => search,
  loader: async ({ context: { queryClient }, deps, cause }) => {
    if (cause !== 'stay') {
      await queryClient.prefetchQuery(queryThatConsumesDeps)
    }
  },
  component: UsersPage,
  pendingComponent: UsersPageSkeleton
})
```
The main idea is that I don't use the built-in router cache, so I pass all the `loaderDeps` to the `loader` and let react-query handle the caching and invalidation. The `cause` property can be `enter`, `stay`, or `preload`. I want the `pendingComponent` to be shown when a user enters the route for the first time (`enter` cause). When the user preloads the route (hover the link, etc), I also want to prefetch queries in the loader. I believe the `stay` cause means that the `loaderDeps` were changed, so in this case, I rely on the loading state from `useQuery`, and do not await anything in loader, so no `pendingComponent` is shown
